### PR TITLE
Try-state for Collective pallet

### PR DESCRIPTION
**Pull request by Szegoo**
This PR writes expectations for the collective pallet storage that must always apply.

Part of: #12831

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2550451

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2550522

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2552031

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2552032

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2552033

**Comment by Szegoo** [comment]
@ggwpez Could you please review this?

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -341,6 +341,14 @@ pub mod pallet {
 		WrongProposalLength,
 	}
 
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		#[cfg(feautre = "try-runtime")]
```
Comment (ggwpez) [inline frame/collective/src/lib.rs:346]:
```suggestion
		#[cfg(feature = "try-runtime")]
```

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -341,6 +341,14 @@ pub mod pallet {
...
+		fn try_state(_n: BlockNumberFor<T>) -> Result<(), &'static str> {
+			Self::do_try_state();
```
Comment (ggwpez) [inline frame/collective/src/lib.rs:348]:
```suggestion
			Self::do_try_state()
```

File: frame/collective/src/lib.rs [hunk 3]
```diff
@@ -968,6 +976,48 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///
+	/// Looking at members:
+	/// * The members count must never exceed `MaxMembers`.
+	#[cfg(any(feature = "try-runtime", test))]
+	fn do_try_state() -> Result<(), &'static str> {
+		Self::proposals().into_iter().for_each(|proposal| {
+			assert!(Self::proposal_of(proposal).is_some());
+		});
+
+		assert!(Self::proposals().into_iter().count() <= Self::proposal_count() as usize);
```
Comment (ggwpez) [inline frame/collective/src/lib.rs:1007]: Rather use `ensure!` here since this function should return an `Err`, not panic.

**Review by ggwpez — COMMENTED** [review]
Generally looks good.  
There are probably more things that can be checked - will take a closer look tomorrow.

File: frame/collective/src/lib.rs (hunk 3 above)
Comment (Szegoo) [reply frame/collective/src/lib.rs:1007]: But aren't we panicking in other `try_state` hooks? Looking at the nomination pools: https://github.com/paritytech/substrate/blob/8bd57418b79d2450079755c0d631168b8fad2e97/frame/nomination-pools/src/lib.rs#L3059-L3072

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2573470

**Comment by Szegoo** [comment]
> There are probably more things that can be checked 

Added two more checks. Please let me know if you come up with some more things that can be checked.

File: frame/collective/src/lib.rs (hunk 3 above)
Comment (ggwpez) [reply frame/collective/src/lib.rs:1011]: Huh, weird. Then why does it return an error @kianenigma :rofl: ?

File: frame/collective/src/lib.rs [hunk 4]
```diff
@@ -968,6 +976,65 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	/// ## Expectations:
+	///
+	/// Looking at proposals:
+	///
+	/// * Each hash of a proposal that is stored inside `Proposals` must have a
+	/// call mapped to it inside the `ProposalsOf` storage map.
+	/// * `ProposalCount` must always be more or equal to the number of
+	/// proposals inside the `Proposals` storage value. The reason why
+	/// `ProposalCount` can be more is because when a proposal is removed the
+	/// count is not deducted.
```
Comment (liamaharon) [inline frame/collective/src/lib.rs:993]: I think another invariant to check is that the count of values in `ProposalsOf` matches the count of values in `Proposals`. Combined with the first check, this implicitly checks that each key (hash) in `ProposalsOf` exists in `Proposals`.

File: frame/collective/src/lib.rs [hunk 5]
```diff
@@ -968,6 +976,65 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	///
+	/// Looking at votes:
+	/// * The sum of aye and nay votes for a proposal can never exceed
+	///  `MaxMembers`.
+	/// * The proposal index inside the `Voting` storage map must be unique.
+	///
+	/// Looking at members:
+	/// * The members count must never exceed `MaxMembers`.
```
Comment (liamaharon) [inline frame/collective/src/lib.rs:1001]: Another invariant to check is that all members are sorted (by value).

File: frame/collective/src/lib.rs (hunk 4 above)
Comment (liamaharon) [reply frame/collective/src/lib.rs:993]: also, the `threshold` of a proposal should always be `> 1`

File: frame/collective/src/lib.rs [hunk 6]
```diff
@@ -968,6 +976,65 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+	/// call mapped to it inside the `ProposalsOf` storage map.
+	/// * `ProposalCount` must always be more or equal to the number of
+	/// proposals inside the `Proposals` storage value. The reason why
+	/// `ProposalCount` can be more is because when a proposal is removed the
+	/// count is not deducted.
+	///
+	/// Looking at votes:
+	/// * The sum of aye and nay votes for a proposal can never exceed
+	///  `MaxMembers`.
+	/// * The proposal index inside the `Voting` storage map must be unique.
```
Comment (liamaharon) [inline frame/collective/src/lib.rs:998]: I think another invariant to check: the vote proposal hash must exist in `Proposals`

**Review by liamaharon — COMMENTED** [review]
Looking good! Added some ideas for additional variants to check.

**Comment by Szegoo** [comment]
@ggwpez @kianenigma If all good could you approve? 🙇

**Comment by Szegoo** [comment]
> Still dont know if we should use `ensure` or `assert` 😆

Yes, that is still a mystery 🤣

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2596700

**Review by liamaharon — APPROVED** [review]

File: frame/collective/src/lib.rs (hunk 3 above)
Comment (kianenigma) [reply frame/collective/src/lib.rs:1014]: I was never happy with the `&'static str` so I may have ignored it in a few places 🙈  not very nice of me. Please return an error and see: https://github.com/paritytech/substrate/issues/13736

**Review by kianenigma — COMMENTED** [review]
Prior to merging, I'd like some evidence that the given expectations actually pass against the current state of Polkadot, Kusama, and Westend. Can you check that?

**Comment by Szegoo** [comment]
> Prior to merging, I'd like some evidence that the given expectations actually pass against the current state of Polkadot, Kusama, and Westend. Can you check that?

Yes, I will check that.

**Comment by Szegoo** [comment]
@kianenigma I have run the try-state against the Polkadot state. AFAIK Kusama and Westend don't use the pallet_collective in their runtime.

**Comment by Szegoo** [comment]
Screenshot 2023-04-14 at 12 06 46

**Comment by kianenigma** [comment]
> [April 16, 2023 15:08 ](#event-9017469853)
> Revert
> View details[@liamaharon](/liamaharon) [liamaharon](/liamaharon) merged commit [`4e93782`](/paritytech/substrate/commit/4e93782c16f741cde18a8b98bd791838e03788fb) into paritytech:master [Apr 17, 2023](https://github.com/paritytech/substrate/pull/13645#event-9020046487)
> 47 checks passed

Why? it is crystal clear to me :D 


https://github.com/paritytech/substrate/issues/13736

@Szegoo @liamaharon any of you fancy it?

**Comment by liamaharon** [comment]
> > Looks good. Still dont know if we should use ensure or assert 😆
> 
> Why? it is crystal clear to me :D
> 
> #13736
> 
> @Szegoo @liamaharon any of you fancy it?

@Szegoo you have first dibs, otherwise I can pick it up next week :)

**Comment by Szegoo** [comment]
> > @Szegoo @liamaharon any of you fancy it?
> 
> @Szegoo you have first dibs, otherwise I can pick it up next week :)

I will happily take it :)

**Comment by kianenigma** [comment]
This was used to found a bug, initiating a tip:

**Comment by Szegoo** [comment]
@kianenigma Is tipping on Kusama still used with the introduction of Open Gov? Seems like the tipping value over here on Kusama is so far 0 KSM :sweat_smile:

**Comment by mordamax** [comment]
OpenGov is WIP still testing on staging, soon will be released. Please use polkadot for now
Thx!

**Comment by Szegoo** [comment]
@mordamax Yeah, I usually use Polkadot, but in my Github profile description, my Kusama address was written(if I recall correctly this was put here because of the fellowship) and Kian just pasted my address from there. Updated my GitHub profile description now so this shouldn't happen in the future again.

**Comment by rzadp** [comment]
Hey everyone,

the OpenGov support in the Tip Bot is now deployed to production.
So we can tip on Kusama now (and continue tipping on Polkadot as well).

**Pull request merged**